common/ethereum/src/eth_contracts/ptokens_router.rs

```diff
@@ -0,0 +1,84 @@
+use std::convert::TryFrom;
+
+use common::{types::Bytes, AppError};
+use common_metadata::{MetadataChainId, METADATA_CHAIN_ID_NUMBER_OF_BYTES};
+use derive_more::Constructor;
+use ethabi::{decode as eth_abi_decode, ParamType as EthAbiParamType, Token as EthAbiToken};
+
+use crate::{EthLog, EthLogExt, EthReceipt};
+
+crate::make_topics!(
+    PTOKENS_ROUTER_METADATA_EVENT_TOPIC => "41954c3bf6e497b17fc12f429900878df830619bbcccb5f61aedc91e6ccc9e64",
+);
+
+#[derive(Debug, Clone, Constructor, Eq, PartialEq, Default)]
+pub struct PTokensRouterMetadataEvent {
+    pub user_data: Bytes,
+    pub origin_chain_id: MetadataChainId,
+    pub origin_address: String,
+    pub destination_chain_id: MetadataChainId,
+    pub destination_address: String,
+}
+
+impl TryFrom<&EthLog> for PTokensRouterMetadataEvent {
+    type Error = AppError;
+
+    fn try_from(log: &EthLog) -> std::result::Result<Self, Self::Error> {
```

Comment:
There's no need for the fully qualified path here:

```suggestion
    fn try_from(log: &EthLog) -> Result<Self, Self::Error> {
```